### ISSUE TITLE
Decode function

### ISSUE THREAD

**Issue by vb000** [nttcslab/msm-mae#3]
Hi,

Really cool work! Is it possible to provide decoding counterparts for `encode` and `encode_lms` functions? Something like `decode_to_lms` or `decode_to_wav`? They could be very useful in debugging the generated embeddings...

**Comment by daisukelab** [comment]
Hi @vb000,

Thank you for your interest and comment. Sorry for taking long to reply.
I am adding two new helper functions to the runtime module `RuntimeMAE`, which fulfills your suggestion.

```python
    def reconstruct(self, lms, mask_ratio, start_frame=0):
        """A helper function to get reconstruction results.
        Use `lms_to_wav` if you may also want to convert the reconstruction results to wavs.
        **Note** this does *not* process the entire LMS frames but rather crops them from the start_frame with the duration of the model's unit frame.
        """

        # trim frames
        unit_frames = self.backbone.patch_embed.img_size[1]
        last_frame = start_frame + unit_frames
        lms_cropped = lms[..., start_frame:last_frame]
        # raw reconstruction
        with torch.no_grad():
            loss, recons, errormap, mask = self.backbone.forward_viz(lms_cropped, mask_ratio)

        return loss, lms_cropped, recons, errormap, mask


    def lms_to_wav(self, single_lms, norm_stats, sr=16000, n_fft=400, hop_length=160, win_length=400):
        """A helper function to revert an LMS into an audio waveform.
        Be sure to use the normalization statistics you used to normalize the LMS.
        """

        mu, sigma = norm_stats
        M = (single_lms*sigma + mu).exp().numpy()
        wav = librosa.feature.inverse.mel_to_audio(M, sr=sr, n_fft=n_fft, hop_length=hop_length, win_length=win_length)
        # display(Audio(wav, rate=sr))
        return wav
```

I also summarized an example in [misc/Note_viz_and_play_reconstruction2.ipynb](misc/Note_viz_and_play_reconstruction2.ipynb).
![image](https://user-images.githubusercontent.com/14831220/216757961-71d017dc-f98a-422e-a8ce-fc6582b99837.png)

I think these functions will be helpful for debugging purposes. However, I welcome further suggestions!
Please find my last commit #[e411b0f](https://github.com/nttcslab/msm-mae/commit/e411b0ff1c729ba0fa9ad6089dadb1e3c1eec704).

Thank you very much!

**Comment by vb000** [comment]
Hi @daisukelab, thanks for the response! I realize my question might have been unclear... I was wondering if there is a decoding function that decodes the generated embeddings to waveforms, for the purpose of qualitative verification; in contrast to the existing function in the repo that converts audio back to audio.

The use case I was referring to: if one uses your model to generate audio embeddings over a large dataset, how do we qualitatively make sure the embeddings we computed are correct? If there was a decoding function that can convert the embeddings back to audio, we can listen to them and make sure the embedding makes sense. Reconstructing audio back to audio using your model is helpful to check the reconstruction quality, but as the user of your model, I'd ideally want to qualitatively verify the embeddings generated by the model, as those are the final outputs of your model..

**Comment by daisukelab** [comment]
Hi @vb000, thanks again for the detail of the need. That sounds very reasonable.
I added the decode_to_lms function to reconstruct LMS from the embeddings, which supports the downstream scenario where the entire input is encoded. Unlike my expectation that reverting to the raw audio would sound quite bad, it sounds good enough to understand how the content was.
![image](https://user-images.githubusercontent.com/14831220/216871780-06532b65-3fe4-4697-adca-8e7e58733c8e.png)

Please note that to be very strict, we cannot guarantee that the decoder can reconstruct visible patch embeddings to the original LMS space because the training does not calculate the loss on the reconstruction result of the visible patches. Since MAE calculates only the loss on the masked tokens, the decoder learns to predict the original input patches of the masked tokens using the visible patch tokens.

Please find the commit [947aeee](https://github.com/nttcslab/msm-mae/commit/947aeee845c27e735f698f3c706d9109525ca637). Thanks again!

**Comment by vb000** [comment]
Hi @daisukelab, thanks for the quick response! This is a great addition! It totally makes sense that the reconstruction might not be very clear.
